### ISSUE TITLE
Child item is deleted when moved from one parent collection to another

### ISSUE THREAD

**Issue by johnkwaters** [dotnet/efcore#24787]
I have a Stop object with a collection of Activities
If I move an Activity from on Stop to Another, it get's marked as deleted.
If I just change the Activity.Stop (or Activity.StopId) property of the Activity, it stays in the wrong parent collection (stays in the source Stop.Activities, doesn't move to target Stop.Activities.
So I manually move it: 1) sourceStop.Activities.Remove(activity) and 2) targetStop.Activities.Add(activity)
Now all is well... except when I save the changes... the activity is DELETED!
Why? Because the removing it from the sourceStop.Activities marks it's state as Deleted.

This seems to be a bug?

How do you move a child from the collection of one parent to another, and have not only the navigation property update, but also the collections themselves update?

**Comment by johnkwaters** [comment]
`       /// <summary>
        /// Copy or move a ticket to a new stop. If the ticket has any survey results, copy it
        /// (to preserve the survey result), if not, just moved it
        /// </summary>
        /// <param name="srcTicket">The ticket to copy or move</param>
        /// <param name="targetStop">The stop to move or copy it to</param>
        /// <returns>The original ticket moved, or new one if copied</returns>
        public async Task<Activity> CopyOrMoveTicketAsync(Activity srcTicket, Stop targetStop)
        {
            // we need to save this value here to preserve it, as it is nulled out
            // EF if ticket is assigned to a new stop, but the original value is needed
            var srcStop = srcTicket.Stop;

            if (targetStop.Activities == null)
            {
                targetStop.Activities = new List<Activity>();
            }

            // if the ticket has a survey results, copy to a new ticket, otherwise just
            // move it.
            if (srcTicket.SurveyResultId.HasValue)
            {
                // create a copy of the activity (with properties)
                var newActivity = _mapper.Map<Activity>(srcTicket);
                newActivity.SurveyResultId = null;
                await _activityRepo.CreateAsync(newActivity);

                targetStop.Activities.Add(newActivity);

                // change the old ticket external id to be different
                if (!string.IsNullOrEmpty(srcTicket.ExternalId))
                {
                    srcTicket.ExternalId += '*';
                }
                newActivity.ActivityStatus = ActivityStatus.Open;
                return newActivity;
            }

            // we can just move it
            srcTicket.Stop = targetStop;

            // make sure it is moved from the old stop
            if (srcStop.Activities != null && srcStop.Activities.Any() && srcStop.Activities.Contains(srcTicket))
            {
                srcStop.Activities.Remove(srcTicket);
            }
            // make sure it has been added to the new stop
            if (!targetStop.Activities.Contains(srcTicket))
            {
                targetStop.Activities.Add(srcTicket);
            }
            // make sure the 'remove' above didn't delete the object!
            _tripRepo.Context.Entry(srcTicket).State = EntityState.Modified;

            if (ShouldDeleteStopIfEmpty(srcStop))
            {
                await DeleteStopAsync(srcStop);
            }
            srcTicket.ActivityStatus = ActivityStatus.Open;
            return srcTicket;
        }`

**Comment by johnkwaters** [comment]
See code above and comment

 // make sure the 'remove' above didn't delete the object!
_tripRepo.Context.Entry(srcTicket).State = EntityState.Modified;

That's my workaround for the bug - there must be a better way?

**Comment by ajcvickers** [comment]
Please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by johnkwaters** [comment]
@ajcvickers The code I pasted above can't really be extracted from my project and really I don't know if it's even possible to create a small,. runnable project that uses SQL Server? I mean I would have to define the master and detail tables, populate them with data, write code to fetch and move data around etc... doesn't sound feasible. Can you not tell from the code above? It's pretty straight forward: an item is in a child collection of object 1 and I want it to move to object 2's collection.

**Comment by ajcvickers** [comment]
@johnkwaters No, I can't tell from the code, sorry. I need to understand what the model is, how the context is configured, and how the context is created and used outside of the code you show.

This is the template I usually use for creating repros:

```C#
public class Foo
{
    public int Id { get; set; }
}

public class SomeDbContext : DbContext
{
    private static ILoggerFactory ContextLoggerFactory
        => LoggerFactory.Create(b => b.AddConsole().SetMinimumLevel(LogLevel.Information));

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseSqlServer(Your.ConnectionString)
            .UseLoggerFactory(ContextLoggerFactory)
            .EnableSensitiveDataLogging();

    public DbSet<Foo> Foos { get; set; }
    
    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
    }
}

public class Program
{
    public static void Main()
    {
        using (var context = new SomeDbContext())
        {
            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();

            // Insert some test data here. For example...
            context.Add(new Foo());
            context.SaveChanges();
        }

        using (var context = new SomeDbContext())
        {
            // Write test code here...
        }
    }
}
```

It really shouldn't be too hard to extract your code and model into this kind of template. If it is too hard _for you_, then _we_ have no chance of doing it given we don't have your code to look at. And since I don't see anything obviously wrong, then the investigation is pretty much dead-ended.

**Comment by johnkwaters** [comment]
I am working on it.... coming up

**Comment by johnkwaters** [comment]
So this demonstrates stage 1 of the problem:

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;

namespace ConsoleApp7
{
    class Program
    {
        static async Task Main(string[] args)
        {
            Console.WriteLine("Hello World!");

            using (var context = new PocContext())
            {
                if (!await context.Parents.AnyAsync())
                {
                    await SeedAsync(context);
                }

                var p1 = await context.Parents
                    .Include( p => p.Children)
                    .SingleAsync(p => p.Id == 1);
                var p2 = await context.Parents
                    .Include(p => p.Children)
                    .SingleAsync(p => p.Id == 2);

                var child = p1.Children.First();

                Debug.WriteLine($"BEFORE: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"BEFORE: Child's parent has ID {child.Parent.Id}");

                child.Parent = p2;

                Debug.WriteLine($"AFTER: Child's parent has ID {child.Parent.Id}");
                Debug.WriteLine($"AFTER: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
            }
        }

        private static async Task SeedAsync(PocContext context)
        {
            var parent1 = new Parent()
            {
                Children = new List<Child>()
            };
            var parent2 = new Parent()
            {
            };
            var child1 = new Child()
            {
            };
            parent1.Children.Add((child1));

            context.Parents.Add(parent1);
            context.Parents.Add(parent2);

            await context.SaveChangesAsync();
        }
    }

    public class Parent
    {
        public int Id { get; set; }

        public List<Child> Children { get; set; }
    }

    public class Child
    {
        public Parent Parent { get; set; }
        public int Id { get; set; }
    }

    public class PocContext : DbContext
    {
        public DbSet<Parent> Parents { get; set; }
        public DbSet<Child> Children { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"Server=.;Database=POC;Trusted_Connection=True;");
        }
    }
}

OUTPUT
BEFORE: Parent 1 has ID 1 and 1 children
BEFORE: Child's parent has ID 1
AFTER: Child's parent has ID 2
AFTER: Parent 1 has ID 1 and 1 children

**Comment by johnkwaters** [comment]
As you can see, after moving the child from p1 to p2 by changing it's parent navigational property, it is STILL in p1s child collection...it shouldn't be there IMO 
So that's why I call "remove" on the p1 child collection and "add" on the p2 to move it.. but that first remove deletes the child record.
Trying to reproduce that next.

**Comment by johnkwaters** [comment]
Added the rest:

        static async Task Main(string[] args)
        {
            Console.WriteLine("Hello World!");

            using (var context = new PocContext())
            {
                if (!await context.Parents.AnyAsync())
                {
                    await SeedAsync(context);
                }

                var p1 = await context.Parents
                    .Include( p => p.Children)
                    .SingleAsync(p => p.Id == 1);
                var p2 = await context.Parents
                    .Include(p => p.Children)
                    .SingleAsync(p => p.Id == 2);

                var child = p1.Children.First();

                Debug.WriteLine($"BEFORE: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"BEFORE: Child's parent has ID {child.Parent.Id}");

                child.Parent = p2;

                Debug.WriteLine($"AFTER: Child's parent has ID {child.Parent.Id}");
                Debug.WriteLine($"AFTER: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"AFTER: Parent 2 has ID {p2.Id} and {p2.Children.Count} children");

                p1.Children.Remove(child);
                p2.Children.Add(child);

                Debug.WriteLine($"AFTER 2: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"AFTER 2: Parent 2 has ID {p2.Id} and {p2.Children.Count} children");

                await context.SaveChangesAsync();

                p1 = await context.Parents
                    .Include(p => p.Children)
                    .SingleAsync(p => p.Id == 1);
                p2 = await context.Parents
                    .Include(p => p.Children)
                    .SingleAsync(p => p.Id == 2);

                Debug.WriteLine($"AFTER 3: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"AFTER 3: Parent 2 has ID {p2.Id} and {p2.Children.Count} children");
            }
        }

OUTPUT

BEFORE: Parent 1 has ID 1 and 1 children
BEFORE: Child's parent has ID 1
AFTER: Child's parent has ID 2
AFTER: Parent 1 has ID 1 and 1 children
AFTER: Parent 2 has ID 2 and 0 children
AFTER 2: Parent 1 has ID 1 and 0 children
AFTER 2: Parent 2 has ID 2 and 1 children
AFTER 3: Parent 1 has ID 1 and 0 children
AFTER 3: Parent 2 has ID 2 and 1 children

**Comment by johnkwaters** [comment]
So in this repro the child does NOT get deleted - but it does in my app.

But issue 1 is shouldn't the child move to the other tracked entities child collection?
It doesn't longer belong to p1 after it was parented.
Any code further along that iterates over p1s children isn't expecting to find p2 children in there mixed in, right?
And you can't assume the code can do savechanges immediately on reparenting.

Maybe would calling DetectChanges work? Or is my expectation of it 'moving' just wrong?

**Comment by johnkwaters** [comment]
Also in the demo above:

var state = context.Entry(child).State;

Shows Modified, whereas in my code the state is Deleted.

**Comment by johnkwaters** [comment]
I have found that if I manually call context.ChangeTracker.DetectChanges() after changing the parent of the child, the child DOES move from parent1 to parent2's child collection,

**Comment by johnkwaters** [comment]
Here is the new version

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace ConsoleApp7
{
    class Program
    {
        static async Task Main(string[] args)
        {
            Console.WriteLine("Hello World!");

            using (var context = new PocContext())
            {
                await context.Database.EnsureDeletedAsync();
                await context.Database.EnsureCreatedAsync();

                await SeedAsync(context);

                var p1 = await context.Parents
                    .Include( p => p.Children)
                    .SingleAsync(p => p.Id == 1);
                var p2 = await context.Parents
                    .Include(p => p.Children)
                    .SingleAsync(p => p.Id == 2);

                var child = p1.Children.First();

                Debug.WriteLine($"BEFORE: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"BEFORE: Child's parent has ID {child.Parent.Id}");

                child.Parent = p2;
                var state = context.Entry(child).State;

                Debug.WriteLine($"AFTER: Child's parent has ID {child.Parent.Id}");
                Debug.WriteLine($"AFTER: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"AFTER: Parent 2 has ID {p2.Id} and {p2.Children.Count} children");
                Debug.WriteLine($"AFTER: child state is {state}");

                context.ChangeTracker.DetectChanges();

                state = context.Entry(child).State;

                Debug.WriteLine($"AFTER DetectChanges: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"AFTER DetectChanges: Parent 2 has ID {p2.Id} and {p2.Children.Count} children");
                Debug.WriteLine($"AFTER DetectChanges2: child state is {state}");

                if (p1.Children.Contains(child))
                {
                    // this code doesn't execute if DetectChange is called, otherwise it does
                    p1.Children.Remove(child);
                }

                if (!p2.Children.Contains(child))
                {
                    // this code doesn't execute if DetectChange is called, otherwise it does
                    p2.Children.Add(child);
                }

                await context.SaveChangesAsync();

                p1 = await context.Parents
                    .Include(p => p.Children)
                    .SingleAsync(p => p.Id == 1);
                p2 = await context.Parents
                    .Include(p => p.Children)
                    .SingleAsync(p => p.Id == 2);

                Debug.WriteLine($"AFTER Refetch: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"AFTER Refetch: Parent 2 has ID {p2.Id} and {p2.Children.Count} children");
            }
        }

        private static async Task SeedAsync(PocContext context)
        {
            var parent1 = new Parent()
            {
                Children = new List<Child>()
            };
            var parent2 = new Parent()
            {
            };
            var child1 = new Child()
            {
            };
            parent1.Children.Add((child1));

            context.Parents.Add(parent1);
            context.Parents.Add(parent2);

            await context.SaveChangesAsync();
        }
    }

    public class Parent
    {
        public int Id { get; set; }

        public List<Child> Children { get; set; }
    }

    public class Child
    {
        public Parent Parent { get; set; }
        public int Id { get; set; }
    }

    public class PocContext : DbContext
    {
        private static ILoggerFactory ContextLoggerFactory
            => LoggerFactory.Create(b => b.AddConsole().SetMinimumLevel(LogLevel.Information));

        public DbSet<Parent> Parents { get; set; }
        public DbSet<Child> Children { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"Server=.;Database=POC;Trusted_Connection=True;");
        }
    }
}

OUTPUT

BEFORE: Parent 1 has ID 1 and 1 children
BEFORE: Child's parent has ID 1
AFTER: Child's parent has ID 2
AFTER: Parent 1 has ID 1 and 0 children
AFTER: Parent 2 has ID 2 and 1 children
AFTER: child state is Modified
AFTER DetectChanges: Parent 1 has ID 1 and 0 children
AFTER DetectChanges: Parent 2 has ID 2 and 1 children
AFTER DetectChanges2: child state is Modified
AFTER Refetch: Parent 1 has ID 1 and 0 children
AFTER Refetch: Parent 2 has ID 2 and 1 children

**Comment by johnkwaters** [comment]
I also found that if I just examine the child's state after moving it, this forces it to 'jump' to the new collection!

        static async Task Main(string[] args)
        {
            Console.WriteLine("Hello World!");

            using (var context = new PocContext())
            {
                await context.Database.EnsureDeletedAsync();
                await context.Database.EnsureCreatedAsync();

                await SeedAsync(context);

                var p1 = await context.Parents
                    .Include( p => p.Children)
                    .SingleAsync(p => p.Id == 1);
                var p2 = await context.Parents
                    .Include(p => p.Children)
                    .SingleAsync(p => p.Id == 2);

                var child = p1.Children.First();

                Debug.WriteLine($"BEFORE: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"BEFORE: Child's parent has ID {child.Parent.Id}");

                child.Parent = p2;

                // if I uncomment this line, it moves to the new collection when I check it's state!
                //var state1 = context.Entry(child).State;

                Debug.WriteLine($"AFTER: Child's parent has ID {child.Parent.Id}");
                Debug.WriteLine($"AFTER: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"AFTER: Parent 2 has ID {p2.Id} and {p2.Children.Count} children");
                //Debug.WriteLine($"AFTER: child state is {state1}");

                context.ChangeTracker.DetectChanges();

                var state2 = context.Entry(child).State;

                Debug.WriteLine($"AFTER DetectChanges: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"AFTER DetectChanges: Parent 2 has ID {p2.Id} and {p2.Children.Count} children");
                Debug.WriteLine($"AFTER DetectChanges2: child state is {state2}");

                if (p1.Children.Contains(child))
                {
                    // this code doesn't execute if DetectChange is called, otherwise it does
                    p1.Children.Remove(child);
                }

                if (!p2.Children.Contains(child))
                {
                    // this code doesn't execute if DetectChange is called, otherwise it does
                    p2.Children.Add(child);
                }

                await context.SaveChangesAsync();

                p1 = await context.Parents
                    .Include(p => p.Children)
                    .SingleAsync(p => p.Id == 1);
                p2 = await context.Parents
                    .Include(p => p.Children)
                    .SingleAsync(p => p.Id == 2);

                Debug.WriteLine($"AFTER Refetch: Parent 1 has ID {p1.Id} and {p1.Children.Count} children");
                Debug.WriteLine($"AFTER Refetch: Parent 2 has ID {p2.Id} and {p2.Children.Count} children");
            }
        }

**Comment by ajcvickers** [comment]
@johnkwaters Do any of the examples you posted above reproduce the behavior where the entity is marked as Deleted? I ran the last code you posted with and without DetectChanges and don't see the entity going into a Deleted state. Also. what version of EF Core are you using?

**Comment by johnkwaters** [comment]
No, none of the reproduce the Deleted behavior I am seeing in my code.
However they do show the behavior where after assigning a new parent to a child, it stays in the original parent's child collection.
It moves to the correct place if you check it's state, or call DetectChanges.

So really this bug report is two things

1. Why is it flagged as deleted in my code when I remove it from the original (old) parent's collection) - can't repro here
2. Is this behavior, that is stays in the 'wrong' original child collection expected? It only move when you inspect it's state or call DetectChanges.

Since 2 is a workaround (I call detect changes and it moves, so I don't have to manually call remove on one collection and add on the other), then we could leave it at that and close this.

**Comment by johnkwaters** [comment]
Oh, and I am on the latest EF core 5.05

**Comment by ajcvickers** [comment]
@johnkwaters For 1, I don't know why it is happening and without being able to reproduce it I don't have many ideas for how to proceed.

For 2, this is because EF Core needs to see the change before it can fix up other navigation properties. See [Change detection and notifications](https://docs.microsoft.com/en-us/ef/core/change-tracking/change-detection) for more info.

**Comment by johnkwaters** [comment]
Thanks @ajcvickers - with 2 above I think you have answered that yes, this is expected behavior, and now I know how to make it work correctly. We will have to leave the mystery of the deletion! Thanks for your patience!
